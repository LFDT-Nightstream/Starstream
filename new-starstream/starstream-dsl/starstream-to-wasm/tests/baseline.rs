use starstream_to_wasm::compile;
use starstream_types::{BinaryOp, Block, Expr, Identifier, Literal, Program, Spanned, Statement};

macro_rules! assert_wat_snapshot {
    ($program:expr) => {{
        let program = $program;
        let wasm = compile(&program);
        let mut wat = String::new();
        wasmprinter::Config::new().fold_instructions(true).print(
            &wasm,
            &mut wasmprinter::PrintFmtWrite(&mut wat),
        ).unwrap();
        insta::with_settings!({
            description => format!("{:?}", program),
            omit_expression => true,
            prepend_module_to_snapshot => true,
        }, {
            insta::assert_snapshot!(wat);
        });
    }}
}

#[test]
fn empty() {
    assert_wat_snapshot!(&Program { statements: vec![] });
}

#[test]
fn simple_while_loop() {
    assert_wat_snapshot!(&Program {
        statements: vec![
            Statement::VariableDeclaration {
                name: Identifier::new("foo", None),
                value: Spanned::none(Expr::Literal(Literal::Integer(0))),
            },
            Statement::While {
                condition: Spanned::none(Expr::Binary {
                    op: BinaryOp::Less,
                    left: Box::new(Spanned::none(Expr::Identifier(Identifier::new(
                        "foo", None
                    )))),
                    right: Box::new(Spanned::none(Expr::Literal(Literal::Integer(10)))),
                }),
                body: Block {
                    statements: vec![Statement::Assignment {
                        target: Identifier::new("foo", None),
                        value: Spanned::none(Expr::Binary {
                            op: BinaryOp::Add,
                            left: Box::new(Spanned::none(Expr::Identifier(Identifier::new(
                                "foo", None
                            )))),
                            right: Box::new(Spanned::none(Expr::Literal(Literal::Integer(1)))),
                        }),
                    }],
                },
            },
        ],
    });
}
