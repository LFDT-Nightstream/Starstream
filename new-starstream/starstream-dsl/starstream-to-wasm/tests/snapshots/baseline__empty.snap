---
source: starstream-to-wasm/tests/baseline.rs
description: "Program { statements: [] }"
---
(module
  (type (;0;) (func))
  (export "main" (func 0))
  (func (;0;) (type 0))
)
